Readme.md

# Texty

  light-weight canvas text editing.

## Implementation

  Texty is _not_ an __IDE__, nor a generic rich text editor, it's purpose is to remain light-weight and provide textbox-like editing for canvas. Texty knows nothing of the DOM, and does not interact with the canvas directly, the only interaction is by passing a `CanvasRenderingContext2d` to the `Text#draw(ctx)` method, all caret movements, selection, etc are performed via method calls, the event handlers themselves are designed by the developer however the _example.html_ file contains an example of these basic interactions.

## CSS Styling

 Texty is fully styleable via CSS, performed using the technique mentioned in the article ["Styling canvas drawings with CSS"](http://tjholowaychuk.com/post/6339741902/styling-canvas-drawings-with-css). This even works for inheritance, for example you do not need to define the _font-family_ specifically for the ".text" selector, as the computed style will inherit this.

```css
body {
  font: 14px helvetica, arial, sans-serif;
  padding: 20px;
}
.text {
  font-size: 20px;
  color: #888;
}
.text .line {
  list-style: decimal;
  line-height: 1.2;
}
.text .caret {
  color: red;
}
.text .selection {
  background: rgba(255,0,0,0.1);
}
```

## Build

  To build texty you'll need uglifyjs installed, after which you can simply run:
  
    $ make

  However pre-built scripts are available in _./build_:

     ./build/texty.js
     ./build/texty.min.js

  Texty is roughly 1000 LOC, and when minified is only 24kb.

## Browser support

 Texty is tested in the following browsers but feel free to contribute!
 
   - Chrome 12.x
   - Safari 5.x

## Links

  - try it in your [browser](http://learnboost.github.com/texty/)

## License 

(The MIT License)

Copyright (c) 2011 TJ Holowaychuk &lt;tj@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.0.3 / 2011-09-02 
==================

  * Added text resizing to example
  * Fixed default line-height of null (aka .size)
  * Fixed line-height support
  * Fixed caret sizing

0.0.2 / 2011-07-29 
==================

  * Added `Text#em(ctx)`
  * Added `Text#bounds(ctx)
  * Added `Text#drawBounds(ctx)`
  * Changed: `Text#lineHeight()` with no arg returns value
  * Fixed selection precision by rounding
  * Fixed example checkboxes

0.0.1 / 2011-07-28 
==================

  * Initial release


lib/caret.js

/*!
 * Caret
 * Copyright (c) 2010 LearnBoost <tj@learnboost.com>
 * MIT Licensed
 */

/**
 * Initialize a new `Caret` with the given `text` object.
 *
 * @param {Text} text
 * @api public
 */

Caret = function Caret(text) {
  this.text = text;
  this.pos = 0;
  this.color('#888');
  this.softLeft = /([A-Z][a-z0-9]*|[a-z0-9]+|_+| +)$/;
  this.softRight = /^([A-Z][a-z0-9]*|[a-z0-9]+|_+| +)/;
  this.hardLeft = /(\w+| +)$/;
  this.hardRight = /^(\w+| +)/;
  this.show();
};

/**
 * Show the caret.
 *
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.show = function(){
  this.visible = true;
  return this;
};

/**
 * Hide the caret.
 *
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.hide = function(){
  this.visible = false;
  return this;
};

/**
 * Set the caret `color`.
 *
 * @param {String} color
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.color = function(color){
  if (!color) return this;
  this._color = color;
  return this;
};

/**
 * Move the caret by `n`, relative to
 * the current position.
 *
 * @param {Number} n
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.move = function(n){
  var len = this.text._text.length;
  this.pos += n;
  if (this.pos < 0) this.pos = 0
  else if (this.pos > len) this.pos = len;
  return this;
};

/**
 * Move the caret left by `n` or 1.
 *
 * @param {Number} n
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.moveLeft = function(n){
  return this.move(-(n || 1));
};

/**
 * Move the caret right by `n` or 1.
 *
 * @param {Number} n
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.moveRight = function(n){
  return this.move(n || 1);
};

/**
 * Move the caret up a line, attempting
 * to maintain the current column.
 *
 * @param {Number} n
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.moveUp = function(){
  var lines = this.text._text.slice(0, this.pos).split('\n')
    , curr = lines.pop()
    , prev = lines.pop() || ''
    , n = prev.slice(curr.length).length;
  return this.moveLeft(curr.length + n + 1);
};

/**
 * Move the caret down a line, attempting
 * to maintain the current column.
 *
 * @param {Number} n
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.moveDown = function(){
  var lines = this.text._text.slice(this.pos).split('\n')
    , curr = lines.shift()
    , next = lines.shift() || ''
    , n = next.slice(curr.length).length;
  return this.moveRight(curr.length + n + 1);
};

/**
 * Move the caret relative to the point (`x`, `y`).
 *
 * @param {Number} x
 * @param {Number} y
 * @return {Caret} for chaining
 * @api public
 */

Caret.prototype.moveTo = function(x, y){
  if (2 == arguments.length) {
    this.moveTo(this.text.indexOf(x, y));
  } else {
    this.pos = x;
  }
  return this;
};

/**
 * Move splitting on `regexp` and `direction`,
 * returning the number of chars
 * jumped, which is at _least_ `1`.
 *
 * @param {RegExp} regexp
 * @param {Number} direction
 * @return {Number}
 * @api public
 */

Caret.prototype.moveWithPattern = function(regexp, direction){
  var text = this.text._text
    , captures
    , n = 1;

  if (-1 == direction) {
    text = text.slice(0, this.pos);
  } else {
    text = text.slice(this.pos);
  }

  if (captures = regexp.exec(text)) n = captures[1].length;

  this.move(-1 == direction ? -n : n);

  return n;
};

/**
 * Perform a "soft" move left,
 * returning the number of chars
 * advanced.
 *
 * @return {Number}
 * @api public
 */

Caret.prototype.moveLeftSoft = function(){
  return this.moveWithPattern(this.softLeft, -1);
};

/**
 * Perform a "soft" move right,
 * returning the number of chars
 * advanced.
 *
 * @return {Number}
 * @api public
 */

Caret.prototype.moveRightSoft = function(){
  return this.moveWithPattern(this.softRight, 1);
};

/**
 * Perform a "hard" move left,
 * returning the number of chars
 * advanced.
 *
 * @return {Number}
 * @api public
 */

Caret.prototype.moveLeftHard = function(){
  return this.moveWithPattern(this.hardLeft, -1);
};

/**
 * Perform a "hard" move right,
 * returning the number of chars
 * advanced.
 *
 * @return {Number}
 * @api public
 */

Caret.prototype.moveRightHard = function(){
  return this.moveWithPattern(this.hardRight, 1);
};


lib/style.js

/*!
 * style
 * Copyright (c) 2010 LearnBoost <tj@learnboost.com>
 * MIT Licensed
 */

/**
 * Grab the computed style for `selector`'s `prop`.
 *
 * @param {String} selector
 * @param {String} prop
 * @return {String}
 * @api private
 */

selectorStyle = function style(selector, prop) {
  var cache = style.cache = style.cache || {}
    , cid = selector + ':' + prop;

  if (cache[cid]) return cache[cid];

  var parts = selector.split(/ +/)
    , len = parts.length
    , parent = root = document.createElement('div')
    , child
    , part;

  for (var i = 0; i < len; ++i) {
    part = parts[i];
    child = document.createElement('div');
    parent.appendChild(child);
    parent = child;
    if ('#' == part[0]) {
      child.setAttribute('id', part.substr(1));
    } else if ('.' == part[0]) {
      child.setAttribute('class', part.substr(1));
    }
  }

  document.body.appendChild(root);
  var ret = getComputedStyle(child)[prop];
  document.body.removeChild(root);
  return cache[cid] = ret;
};

lib/emitter.js

/*!
 * EventEmitter
 * Copyright (c) 2010 LearnBoost <tj@learnboost.com>
 * MIT Licensed
 */

/**
 * Convert arguments to an array.
 *
 * @return {Array}
 * @api public
 */

function toArray(args) {
  var ret = [];
  for (var i = 0, len = args.length; i < len; ++i) {
    ret.push(args[i]);
  }
  return ret;
}

/**
 * Initialize a new `EventEmitter`.
 */

EventEmitter = function EventEmitter() {
  this.callbacks = {};
};

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 */

EventEmitter.prototype.on = function(event, fn){
  (this.callbacks[event] = this.callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 */

EventEmitter.prototype.emit = function(event){
  var args = toArray(arguments).slice(1)
    , callbacks = this.callbacks[event];

  if (callbacks) {
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args)
    }
  }

  return this;
};


lib/texty.js

/*!
 * Texty
 * Copyright (c) 2010 LearnBoost <tj@learnboost.com>
 * MIT Licensed
 */

// TODO: cleanup
// TODO: optimize if needed (should be fine)
// TODO: substr -> slice

/**
 * Initialize a new `Text` object
 * with optional `str`.
 *
 * @param {String} str
 * @api public
 */

Text = function Text(str) {
  this.caret = new Caret(this);
  this.hide().text(str || '');
  this.clearCSS();
};

/**
 * Clear CSS inheritance, reverting to the
 * original styling for Texty.
 *
 * @api public
 */

Text.prototype.clearCSS = function(){
  this.caret.color('#888');
  this.listStyle('none')
    .size(20)
    .font('Helvetica')
    .color('#000')
    .selectionBackground('#DFF3FC')
    .lineHeight(null);
};

/**
 * When invoked the text will inherit
 * styles from CSS.
 *
 * @api public
 */

Text.prototype.inheritCSS = function(){
  var s = selectorStyle;

  // text
  this.size(parseInt(s('.text', 'font-size'), 10))
    .font(s('.text', 'font-family'))
    .color(s('.text', 'color'))
    .selectionBackground(s('.text .selection', 'background-color'));

  // lines
  this.listStyle(s('.text .line', 'list-style-type'))
    .lineHeight(parseInt(s('.text .line', 'line-height'), 10));

  // caret
  this.caret.color(s('.text .caret', 'color'));
};

/**
 * Set line-height to `n`.
 *
 * @param {Number} n
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.lineHeight = function(n){
  if (0 == arguments.length) return this._lineHeight || this._size;
  this._lineHeight = n;
  return this;
};

/**
 * Return the string primitive.
 *
 * @return {String}
 * @api public
 */

Text.prototype.toString = function(){
  return this.text();
};

/**
 * Set list `style` to one of:
 *
 *  - `none`
 *  - `decimal` enable line numbering
 *  - `disc` solid circle
 *  - `circle` stroked circle
 *  - `square` solid square
 *
 * @param {String} style
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.listStyle = function(style){
  if (0 == arguments.length) return this._listStyle;
  if (!style) return this;
  this._listStyle = style;
  return this;
};

/**
 * Get or set the text to `str`.
 *
 * @param {String} str
 * @return {Text|String}
 * @api public
 */

Text.prototype.text = function(str){
  if (arguments.length) {
    this._text = str;
    return this;
  } else {
    return this._text;
  }
};

/**
 * Set the text `color`.
 *
 * @param {String} color
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.color = function(color){
  if (!color) return this;
  this._textColor = color;
  return this;
};

/**
 * Set the selection background `color`.
 *
 * @param {String} color
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectionBackground = function(color){
  if (!color) return this;
  this._selectionBackgroundColor = color;
  return this;
};

/**
 * Remove `n` chars relative to the caret position.
 *
 * @param {Number} n
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.remove = function(n){
  return this.removeAt(n, this.caret.pos);
};

/**
 * Remove `n` chars relative to `pos`.
 *
 * @param {Number} n
 * @param {Number} pos
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.removeAt = function(n, pos){
  this._text = this._text.substr(0, pos - n) + this._text.substr(pos);
  return this.caret.move(-n);
};

/**
 * Remove range of characters `from`-`to`.
 *
 * @param {Number} from
 * @param {Number} to
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.removeRange = function(from, to){
  this._text = this._text.substr(0, from) + this._text.substr(to);
  return this.deselect().caret.moveTo(from);
};

/**
 * Return selection object containing:
 *
 *  - `string` the text selected
 *  - `from` the start offset
 *  - `to` the end offset
 *
 * @return {Object}
 * @api public
 */

Text.prototype.selection = function(){
  if (!this.selected) return;
  var from = this.selected.from
    , to = this.selected.to;
  return {
      string: this._text.slice(from, to)
    , from: from
    , to: to
  }
};

/**
 * Remove the text currently selected, throwing
 * an error if no text is selected, otherwise
 * returning the object documented in `Text#selection()`.
 *
 * To check for a selection, use `if (text.selected)`.
 *
 * @return {Object}
 * @api public
 */

Text.prototype.removeSelection = function(){
  if (!this.selected) throw new Error('no selection is made');
  var from = this.selected.from
    , to = this.selected.to;
  return this.removeRange(from, to);
};

/**
 * Remove all characters.
 *
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.clear = function(){
  return this.text('');
};

/**
 * Insert `str` at the curent caret position.
 *
 * @param {String} str
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.insert = function(str){
  return this.insertAt(str, this.caret.pos);
};

/**
 * Insert `str` at `pos`.
 *
 * @param {String} str
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.insertAt = function(str, pos){
  this._text = this._text.substr(0, pos)
    + str + this._text.substr(pos);
  return this.caret.move(str.length);
};

/**
 * Return em for `ctx`.
 *
 * @param {CanvasRenderingContext2D} ctx
 * @return {Number}
 * @api public
 */

Text.prototype.em = function(ctx){
  return ctx.measureText('M').width;
};

/**
 * Return the char index relative to the point (`x`, `y`).
 *
 * @param {Number} x
 * @param {Number} y
 * @return {Number} index
 * @api public
 */

Text.prototype.indexOf = function(x, y){
  var text = this.text()
    , width = this._width
    , size = this.lineHeight()
    , lines = text.split('\n')
    , height = lines.length * size
    , len = lines.length
    , x = x - this.x
    , y = y - this.y
    , ctx = this.ctx
    , pad = 10;

  if (!text.length) return 0;

  // x-axis out of bounds
  if (x < -pad || x > width + pad) return 0;

  // y-axis out of bounds
  if (y < -pad || y > height + pad) return 0;

  // cap y
  y = Math.ceil(Math.min(len, Math.max(1, y / size))) - 1;

  // preceding chars
  var preceding = lines.slice(0, y).join('\n').length;
  if (y) ++preceding;

  // compute index
  var line = lines[y]
    , len = line.length
    , width = 0;

  for (var i = 0; i < len; ++i) {
    width += ctx.measureText(line[i]).width;
    if (width > x) break;
  }

  return i + preceding;
};

/**
 * Select range from point (`x`, `y`) to (`x2`, `y2`).
 *
 * @param {Number} x
 * @param {Number} y
 * @param {Number} x2
 * @param {Number} y2
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.select = function(x, y, x2, y2){
  var from = this.indexOf(x, y)
    , to = this.indexOf(x2, y2); 
  return this.selectRange(from, to).caret.moveTo(x2, y2).text;
};

/**
 * Select word at point (`x`, `y`).
 *
 * @param {Number} x
 * @param {Number} y
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectWordAt = function(x, y){
  var i = this.indexOf(x, y);
  this.caret.moveTo(i);
  var left = this.caret.moveLeftHard();
  this.selectLeft(left);
  this.caret.moveTo(i);
  var right = this.caret.moveRightHard();
  this.selectRight(right);
  this.caret.moveTo(i);
  return this;
};

/**
 * Select `n` chars to the left, initiating
 * the range or adding to the current selection.
 *
 * @param {Number} n
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectLeft = function(n){
  var pos = this.caret.pos
    , n = n || 1
    , selection;

  if (selection = this.selection()) {
    this.selectRange(selection.from - n, selection.to);
  } else {
    this.selectRange(pos, pos + n);
  }

  return this;
};

/**
 * Select `n` chars to the right, initiating
 * the range or adding to the current selection.
 *
 * @param {Number} n
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectRight = function(n){
  var pos = this.caret.pos
    , n = n || 1
    , selection;

  if (selection = this.selection()) {
    this.selectRange(selection.from, selection.to + n);
  } else {
    this.selectRange(pos - n, pos);
  }

  return this;
};

/**
 * Select all the text.
 *
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectAll = function(){
  return this.selectRange(0, this._text.length);
};

/**
 * Select range of text via indexes `from`-`to`.
 *
 * @param {Number} from
 * @param {Number} to
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.selectRange = function(from, to){
  var len = this._text.length
    , tmp;

  if (from > to) tmp = from, from = to, to = tmp;

  this.selected = {
      from: from > 0 ? from : 0
    , to: to < len ? to : len
  };

  return this;
};

/**
 * Deselect any text previously selected, if any.
 *
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.deselect = function(){
  this.selected = null;
  return this;
};

/**
 * Move to point (`x`, `y`).
 *
 * @param {Number} x
 * @param {Number} y
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.moveTo = function(x, y){
  this.x = x;
  this.y = y;
  return this;
};

/**
 * Hide the text.
 *
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.hide = function(){
  this.visible = false;
  return this;
};

/**
 * Show the text.
 *
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.show = function(){
  this.visible = true;
  return this;
};

/**
 * Set the font `families`.
 *
 * @param {String} families
 * @return {Text}
 * @api public
 */

Text.prototype.font = function(families){
  if (0 == arguments.length) return this._families;
  if (!families) return this;
  this._families = families;
  return this;
};

/**
 * Set font size to `n`.
 *
 * @param {Number} n
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.size = function(n){
  if (0 == arguments.length) return this._size;
  if (!n) return this;
  this._size = n;
  return this;
};

/**
 * Draw selection.
 *
 * We split the lines, checking the offset
 * within the first, filling the entire line
 * width for the "middle" lines, and
 * special-casing the "last" selected line
 * to fill the correct offset much like the first.
 *
 * @api private
 */

Text.prototype.drawSelection = function(ctx, x, y, size, text){
  var selection = this.selection()
    , from = selection.from
    , to = selection.to
    , before = text.slice(0, from).split('\n')
    , beforeLen = before.length
    , last = before[beforeLen - 1]
    , selected = text.slice(from, to).split('\n')
    , selectionLen = selected.length
    , lineHeight = this.lineHeight()
    , px = size / 2
    , ox = x
    , oy = y
    , width
    , line;

  ctx.fillStyle = this._selectionBackgroundColor;

  // first
  y += (beforeLen - 1) * lineHeight;
  width = ctx.measureText(selected[0]).width;
  last = ctx.measureText(last).width;
  ctx.fillRect(x + last, y - px, width, size);

  // mid
  for (var i = 1, len = selectionLen - 1; i < len; ++i) {
    y += lineHeight;
    width = ctx.measureText(selected[i]).width;
    ctx.fillRect(x, y - px, width, size);
  }
  
  // last
  if (selectionLen < 2) return;
  y += lineHeight;
  last = selected[selectionLen - 1];
  width = ctx.measureText(last).width;
  ctx.fillRect(x, y - px, width, size);
};

/**
 * Draw the caret.
 *
 * Splits the lines up to the caret pos,
 * allowing us to grab the line containing it,
 * after which we measure text to provide
 * the appropriate offset.
 *
 * @api private
 */

Text.prototype.drawCaret = function(ctx, x, y, size, text){
  if (!this.caret.visible) return;
  var lines = text.substr(0, this.caret.pos).split('\n')
    , lineHeight = this.lineHeight()
    , len = lines.length
    , line = lines[len - 1]
    , px = size / 2
    , caret = ctx.measureText(line).width + 2
    , y = y + --len * lineHeight;
  ctx.strokeStyle = this.caret._color;
  ctx.beginPath();
  ctx.moveTo(x + caret, y - px - 2);
  ctx.lineTo(x + caret, y + px + 2);
  ctx.stroke();
};

/**
 * Draw text.
 *
 * Iterates through the lines, rendering
 * them while progressing the `y` offset.
 *
 * @api private
 */

Text.prototype.drawText = function(ctx, x, y, size, text){
  var lines = text.split('\n')
    , lineHeight = this.lineHeight()
    , ox = x;
  ctx.fillStyle = this._textColor;
  for (var i = 0, len = lines.length; i < len; ++i) {
    ctx.fillText(lines[i], x, y);
    x = ox;
    y += lineHeight;
  }
};

/**
 * Draw line decoration.
 *
 * @api private
 */

Text.prototype.drawLineDecoration = function(ctx, x, y, size, text, c){
  var len = text.split('\n').length
    , lineHeight = this.lineHeight()
    , em = this.em(ctx);

  ctx.save();
  ctx.font = (this._size * .65) + 'px ' + this._families;
  x -= em;
  for (var i = 0; i < len; ++i) {
    ctx.fillText('function' == typeof c
      ? c(i)
      : c, x, y);
    y += lineHeight;
  }
  ctx.restore();
};

/**
 * Return text bounds as an object.
 *
 * @param {CanvasRenderingContext2D} ctx
 * @return {Object}
 * @api public
 */

Text.prototype.bounds = function(ctx){
  var text = this._text
    , lines = text.split('\n')
    , size = Math.max(this._size, this.lineHeight())
    , height = size * lines.length
    , width = 0
    , x = this.x
    , y = this.y;

  for (var i = 0, len = lines.length; i < len; ++i) {
    width = Math.max(width, ctx.measureText(lines[i]).width);
  }

  return {
      x: x
    , y: y - size / 2
    , width: width
    , height: height 
  };
};

/**
 * Draw text boundaries in red.
 *
 * @param {CanvasRenderingContext2D} ctx
 * @return {Text} for chaining
 * @api public
 */

Text.prototype.drawBounds = function(ctx){
  if (!this.visible) return this;
  var bounds = this.bounds(ctx);
  ctx.strokeStyle = 'rgba(255,0,0,.5)';
  ctx.strokeRect(bounds.x, bounds.y, bounds.width, bounds.height);
  return this;
};

/**
 * Draw text, the caret, and possible selection to the given `ctx`.
 *
 * @param {CanvasRenderingContext2D} ctx
 * @return {Text} text
 * @api public
 */

Text.prototype.draw = function(ctx){
  if (!this.visible) return this;
  var text = this._text
    , size = this._size
    , x = this.x
    , y = this.y;

  this.ctx = ctx;
  this._width = ctx.measureText(text).width;
  ctx.textBaseline = 'middle';
  ctx.font = size + 'px ' + this._families;
  if (this.selected) this.drawSelection(ctx, x, y, size, text);
  this.drawCaret(ctx, x, y, size, text);
  this.drawText(ctx, x, y, size, text);

  // list style
  var decoration;
  switch (this._listStyle) {
    case 'decimal': decoration = function(i){ return ++i + '.'; }; break;
    case 'square': decoration = '■'; break;
    case 'circle': decoration = '○'; break;
    case 'disc'  : decoration = '●'; break;
  }
  if (decoration) this.drawLineDecoration(ctx, x, y, size, text, decoration);

  return this;
};
